blockchain/merkle.go

```diff
@@ -154,6 +154,36 @@ func BuildMerkleTreeStore(transactions []*btcutil.Tx, witness bool) []*chainhash
 	return merkles
 }
 
+// CalcMerkleRoot computes the merkle root over a set of hashed leaves. The
+// interior nodes are computed opportunistically as the leaves are added to the
+// abstract tree to reduce the total number of allocations. Throughout the
+// computation, this computation only requires storing O(log n) interior
+// nodes.
+//
+// This method differs from BuildMerkleTreeStore in that the interior nodes are
+// discarded instead of being returned along with the root. CalcMerkleRoot is
+// slightly faster than BuildMerkleTreeStore and requires significantly less
+// memory and fewer allocations.
+//
+// A merkle tree is a tree in which every non-leaf node is the hash of its
+// children nodes. A diagram depicting how this works for bitcoin transactions
+// where h(x) is a double sha256 follows:
+//
+//	         root = h1234 = h(h12 + h34)
+//	        /                           \
+//	  h12 = h(h1 + h2)            h34 = h(h3 + h4)
+//	   /            \              /            \
+//	h1 = h(tx1)  h2 = h(tx2)    h3 = h(tx3)  h4 = h(tx4)
+//
+// The additional bool parameter indicates if we are generating the merkle tree
+// using witness transaction id's rather than regular transaction id's. This
+// also presents an additional case wherein the wtxid of the coinbase transaction
+// is the zeroHash.
+func CalcMerkleRoot(transactions []*btcutil.Tx, witness bool) chainhash.Hash {
+	s := newRollingMerkleTreeStore(uint64(len(transactions)))
```

Comment:
I think you should also do the transformation from []Tx to []Hash here to preserve the API. If you are concerned about allocations you can make the `calcMerkleRoot` function accept []Hashable or something to that effect.

Reply:
I do not believe the CVE you cite is relevant here. There is nothing that ensures the uniqueness of the Tx values in the slice being inputted into this function so nothing is actually prevented by this choice of API. Conflating security checks and data structure computations is not something I would recommend. If you need to do security checks, then do them, but don't make it part of the data structure API.

Reply:
> CalcMerkleRoot is only relevant for merkle root that's committed to the block header and should not be used for anything else because of the vulnerability.

My point here is that your choice to use a Tx slice doesn't actually defend against the vuln. You can simply duplicate the last Tx at the end of the slice and trigger the issue. I looked over this yesterday and didn't see any operation in this implementation where you assert the uniqueness of every transaction in the slice, which is what would be required to sidestep the vulnerability you cite. That tells me that this check has to be done by the caller, and in that case the Tx slice doesn't buy you anything other than an intentionally more constrained API.

> Is there any use case that you have in mind that would require a slice of hashes instead of slice of txs?

in the context of btcd? no. But my point here has always been that this is a broadly useful piece of machinery and Merkle Trees are very commonly used throughout the ecosystem.

If the *only* purpose of this PR is to make btcd block sync fast, then it's probably fine as is. I just think we should take a little bit longer of a view here.

Reply:
> other than an intentionally more constrained API.

Yes. That's desirable.

> But my point here has always been that this is a broadly useful piece of machinery and Merkle Trees are very commonly used throughout the ecosystem.

This specific implementation is not. No one should be using this. Satoshi coded this in so we're stuck with it.

Reply:
> Yes. That's desirable.

I don't understand how a more constrained API, where you don't perform the security checks required to actually dodge the CVE you cite, is desirable.

> This specific implementation is not. No one should be using this. Satoshi coded this in so we're stuck with it.

My understanding of the purpose of this implementation is to compute merkle roots in O(log(n)) memory instead of O(n). What am I missing here? I don't understand what you mean by "Satoshi coded *this* in", what is "this"?

Reply:
FYI my comment here is non-blocking

Reply:
Won't be addressing this in this PR.